cpp/src/arrow/ipc/adapter.cc

```diff
@@ -0,0 +1,296 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/ipc/adapter.h"
+
+#include <cstdint>
+#include <cstring>
+#include <vector>
+
+#include "arrow/array.h"
+#include "arrow/ipc/memory.h"
+#include "arrow/ipc/Message_generated.h"
+#include "arrow/ipc/metadata.h"
+#include "arrow/ipc/metadata-internal.h"
+#include "arrow/schema.h"
+#include "arrow/table.h"
+#include "arrow/type.h"
+#include "arrow/types/construct.h"
+#include "arrow/types/primitive.h"
+#include "arrow/util/buffer.h"
+#include "arrow/util/status.h"
+
+namespace arrow {
+namespace ipc {
+
+static bool IsPrimitive(const DataType* type) {
+  switch (type->type) {
+    case Type::NA:
+      return false;
+    case Type::BOOL:
+    case Type::UINT8:
+    case Type::INT8:
+    case Type::UINT16:
+    case Type::INT16:
+    case Type::UINT32:
+    case Type::INT32:
+    case Type::UINT64:
+    case Type::INT64:
+    case Type::FLOAT:
+    case Type::DOUBLE:
+      return true;
+    default:
+      return false;
+  }
+}
+
+// ----------------------------------------------------------------------
+// Row batch write path
+
+Status VisitArray(const Array* arr, std::vector<flatbuf::FieldNode>* field_nodes,
+    std::vector<std::shared_ptr<Buffer>>* buffers) {
+  if (IsPrimitive(arr->type().get())) {
+    const PrimitiveArray* prim_arr = static_cast<const PrimitiveArray*>(arr);
+
+    field_nodes->push_back(flatbuf::FieldNode(prim_arr->length(),
+            prim_arr->null_count()));
```

Comment:
I don't think the spacing is consistent, not going to knit picking on formatting but I think we should adhere to some style, like the Google C++ style.